### PR TITLE
[Vectorized][java-udf] add datetime&&largeint&&decimal type to java-udf

### LINKED ISSUE #9439: [Vectorized][java-udf] add datetime&&largeint&&decimal type to java-udf
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

support more datatype to java-udf

Date: LocalDate
DateTime: LocalDateTime
LargeInt: BigInteger
Decimal: BigDecimal

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by zhangstar333**
# Proposed changes

Issue Number: #9439 #8389

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/util/jni-util.cpp [hunk 1]
```diff
@@ -16,6 +16,7 @@
 // under the License.
 
 #include "util/jni-util.h"
+#include "jni_md.h"
```
Comment (Gabriel39) [inline be/src/util/jni-util.cpp:19]: move this line below #ifdef LIBJVM

File: be/src/util/jni-util.cpp [hunk 2]
```diff
@@ -45,8 +46,10 @@ void FindOrCreateJavaVM() {
         vm_args.nOptions = 1;
         vm_args.ignoreUnrecognized = JNI_TRUE;
 
-        int res = JNI_CreateJavaVM(&g_vm, (void**)&env, &vm_args);
-        DCHECK_LT(res, 0) << "Failed tp create JVM, code= " << res;
+        jint res = JNI_CreateJavaVM(&g_vm, (void**)&env, &vm_args);
+        if (JNI_OK != res) {
+            DCHECK(false) << "Failed tp create JVM, code= " << res;
```
Comment (Gabriel39) [inline be/src/util/jni-util.cpp:51]:
```suggestion
            DCHECK(false) << "Failed to create JVM, code= " << res;
```

File: be/src/vec/functions/function_java_udf.cpp [hunk 3]
```diff
@@ -49,10 +49,12 @@ JavaFunctionCall::JavaFunctionCall(const TFunction& fn, const DataTypes& argumen
 
 Status JavaFunctionCall::prepare(FunctionContext* context,
                                  FunctionContext::FunctionStateScope scope) {
-    DCHECK(executor_cl_ == NULL) << "Init() already called!";
-    JNIEnv* env;
+    //DCHECK(executor_cl_ == NULL) << "Init() already called!";
```
Comment (Gabriel39) [inline be/src/vec/functions/function_java_udf.cpp:52]: maybe remove it directly

**Review by Gabriel39 — CHANGES_REQUESTED** [review]

File: fe/java-udf/src/main/java/org/apache/doris/udf/UdfExecutor.java [hunk 4]
```diff
@@ -307,50 +326,111 @@ private boolean storeUdfResult(Object obj, long row) throws UdfRuntimeException
         switch (retType_) {
             case BOOLEAN: {
                 boolean val = (boolean) obj;
-                UdfUtils.UNSAFE.putByte(UdfUtils.UNSAFE.getLong(null, outputBufferPtr_) + row * retType_.getLen(), val ? (byte) 1 : 0);
```
Comment (Gabriel39) [inline fe/java-udf/src/main/java/org/apache/doris/udf/UdfExecutor.java:310]: Does this line have a format problem?

File: fe/java-udf/src/main/java/org/apache/doris/udf/UdfExecutor.java [hunk 5]
```diff
@@ -365,41 +445,81 @@ private void allocateInputObjects(long row) throws UdfRuntimeException {
...
+                case LARGEINT: {
+                    long base =
+                            UdfUtils.UNSAFE.getLong(null, UdfUtils.getAddressAtOffset(inputBufferPtrs_, i)) + 16L * row;
+                    byte[] bytes = new byte[16];
+                    UdfUtils.copyMemory(null, base, bytes, UdfUtils.BYTE_ARRAY_OFFSET, 16);
+
+                    inputObjects_[i] = new BigInteger(convertByteOrder(bytes));
+                    break;
+                }
+                case DECIMALV2: {
```
Comment (Gabriel39) [inline fe/java-udf/src/main/java/org/apache/doris/udf/UdfExecutor.java:496]: better to add some UTs in org.apache.doris.udf.UdfExecutorTest for these complex types.

File: fe/java-udf/src/main/java/org/apache/doris/udf/UdfExecutor.java [hunk 6]
```diff
@@ -307,50 +324,113 @@ private boolean storeUdfResult(Object obj, long row) throws UdfRuntimeException
...
+                UdfUtils.UNSAFE.putLong(UdfUtils.UNSAFE.getLong(null, outputBufferPtr_) + row * retType_.getLen(),
+                        time);
+                return true;
+            }
+            case LARGEINT: {
+                BigInteger data = (BigInteger) obj;
+                byte[] bytes = convertByteOrder(data.toByteArray());
+
+                //here  value is 16 bytes, so if result data greater than the maximum of 16 bytes
+                //it will return a wrong num to backend;
```
Comment (Gabriel39) [inline fe/java-udf/src/main/java/org/apache/doris/udf/UdfExecutor.java:385]: If a decimal scale is beyond our limit in BE, we should log a warning and just return NULL.

**Comment by Gabriel39** [comment]
LGTM

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**